File: csw-time-client/src/test/scala/csw/time/client/TimeServiceTest.scala
package csw.time.client

import java.time._

import akka.actor.ActorSystem
import akka.testkit.TestProbe
import csw.time.api.models.CswInstant.TaiInstant
import csw.time.api.scaladsl.TimeService
import csw.time.client.extensions.RichInstant.RichInstant
import org.scalatest.concurrent.Eventually
import org.scalatest.time.SpanSugar.convertDoubleToGrainOfTime
import org.scalatest.{BeforeAndAfterAll, FunSuite, Matchers}

import scala.concurrent.Await

class TimeServiceTest extends FunSuite with Matchers with BeforeAndAfterAll with Eventually {
  private val TaiOffset = 37 // At the time of writing this, TAI is ahead of UTC by 37 seconds.

  override implicit val patienceConfig: PatienceConfig = PatienceConfig(50.millis, 10.millis)
  private implicit val system: ActorSystem             = ActorSystem("time-service")
  private val timeService: TimeService                 = TimeServiceFactory.make(TaiOffset)

  private val testProperties: TestProperties = TestProperties.instance
  import testProperties._

  override protected def afterAll(): Unit = Await.result(system.terminate(), 5.seconds)

  //------------------------------UTC-------------------------------

  //DEOPSCSW-532: Synchronize activities with other comp. using UTC
  //DEOPSCSW-533: Access parts of UTC date.time in Java and Scala
  test("should get UTC time") {
    val utcInstant            = timeService.utcTime()
    val fixedInstant: Instant = Instant.now()

    val expectedMillis = fixedInstant.toEpochMilli +- 5

    utcInstant.value.toEpochMilli shouldEqual expectedMillis
  }

  //DEOPSCSW-534: PTP accuracy and precision while reading UTC
  test("should get precision up to nanoseconds in UTC time") {
    // This test is written in eventually block because coincidentally digits at nano place can be "000".
    // To cover that edge-case eventually is used.
    // formatNanos formats instant in 9 digits precision and if clock does not support nano precision (supports us precision)
    // then last 3 digits will be printed as 000 (This test will fail with jdk's Instant.now call which does not provide nano precision)

    eventually(timeService.utcTime().value.formatNanos(precision) should not endWith "000")
  }

  //DEOPSCSW-537: Optimum way for conversion from UTC to TAI
  test("should convert UTC time to TAI time") {
    val utcInstant = timeService.utcTime()

    val taiInstant = timeService.toTai(utcInstant)

    Duration.between(utcInstant.value, taiInstant.value).getSeconds shouldBe TaiOffset
  }

  //------------------------------TAI-------------------------------

  //DEOPSCSW-535: Synchronize activities with other comp, using TAI
  //DEOPSCSW-536: Access parts of TAI date/time in Java and Scala
  //DEOPSCSW-530: SPIKE: Get TAI offset and convert to UTC and Vice Versa
  test("should get TAI time") {
    val taiInstant          = timeService.taiTime()
    val TAIInstant: Instant = Instant.now().plusSeconds(TaiOffset)

    val expectedMillis = TAIInstant.toEpochMilli +- 5

    taiInstant.value.toEpochMilli shouldEqual expectedMillis
  }

  //DEOPSCSW-538: PTP accuracy and precision while reading TAI
  test("should get precision up to nanoseconds in TAI time") {
    // This test is written in eventually block because coincidentally digits at nano place can be "000".
    // To cover that edge-case eventually is used.

    eventually(timeService.taiTime().value.formatNanos(precision) should not endWith "000")
  }

  //DEOPSCSW-530: SPIKE: Get TAI offset and convert to UTC and Vice Versa
  test("should get TAI offset") {
    val offset = timeService.taiOffset()
    offset shouldEqual TaiOffset
  }

  //DEOPSCSW-537: Optimum way for conversion from UTC to TAI
  test("should convert TAI time to UTC time") {
    val taiInstant = timeService.taiTime()

    val utcInstant = timeService.toUtc(taiInstant)

    Duration.between(utcInstant.value, taiInstant.value).getSeconds shouldBe TaiOffset
  }

  //------------------------------Scheduling-------------------------------

  //DEOPSCSW-542: Schedule a task to execute in future
  test("should schedule a task once at given start time with allowed jitter") {
    val testProbe = TestProbe() //todo: Replace with promise

    val idealScheduleTime: TaiInstant = TaiInstant(timeService.taiTime().value.plusSeconds(1))

    timeService.scheduleOnce(idealScheduleTime) {
      testProbe.ref ! timeService.taiTime()
    }

    val actualScheduleTime: TaiInstant = testProbe.expectMsgType[TaiInstant]

    actualScheduleTime.value.getEpochSecond - idealScheduleTime.value.getEpochSecond shouldBe 0
    actualScheduleTime.value.getNano - idealScheduleTime.value.getNano should be < allowedJitterInNanos
  }
}

File: csw-time-client/src/test/java/csw/time/client/TimeServiceTest.java
package csw.time.client;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.testkit.typed.javadsl.ManualTime;
import akka.actor.testkit.typed.javadsl.TestKitJunitResource;
import akka.actor.typed.internal.adapter.ActorSystemAdapter;
import akka.testkit.TestProbe;
import csw.time.api.models.Cancellable;
import csw.time.api.models.TMTTime.TAITime;
import csw.time.api.models.TMTTime.UTCTime;
import csw.time.api.TimeService;
import org.junit.Rule;
import org.junit.Test;
import org.scalatest.junit.JUnitSuite;
import scala.concurrent.duration.FiniteDuration;

import java.time.Duration;
import java.time.Instant;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class TimeServiceTest extends JUnitSuite {
    @Rule
    public TestKitJunitResource testKit = new TestKitJunitResource(ManualTime.config());

    private static int TaiOffset = 37;
    private TestProperties testProperties = TestProperties$.MODULE$.instance();
    private ActorSystem untypedSystem = ActorSystemAdapter.toUntyped(testKit.system());
    private ManualTime manualTime = ManualTime.get(testKit.system());

    private TimeService timeService = TimeServiceFactory.make(TaiOffset, untypedSystem);

    //------------------------------UTC-------------------------------

    //DEOPSCSW-533: Access parts of UTC date.time in Java and Scala
    @Test
    public void should_get_utc_time() {
        UTCTime utcTime = timeService.utcTime();
        Instant fixedInstant = Instant.now();

        long expectedMillis = fixedInstant.toEpochMilli();

        assertEquals((double) expectedMillis, (double) utcTime.value().toEpochMilli(), 5);
    }

    //DEOPSCSW-537: Scala and Java API for conversion between TAI and UTC
    //DEOPSCSW-530: SPIKE: Get TAI offset and convert to UTC and Vice Versa
    @Test
    public void should_convert_utc_to_tai() {
        UTCTime utcTime = timeService.utcTime();
        TAITime taiTime = timeService.toTAI(utcTime);

        assertEquals(Duration.between(utcTime.value(), taiTime.value()).getSeconds(), TaiOffset);
    }

    //DEOPSCSW-534: PTP accuracy and precision while reading UTC
    @Test
    public void should_get_maximum_precision_supported_by_system_in_utc() {
        assertFalse(TestUtil.formatWithPrecision(timeService.utcTime().value(), testProperties.precision()).endsWith("000"));
    }

    //DEOPSCSW-541: PTP accuracy and precision while reading remote location time
    @Test
    public void should_get_maximum_precision_while_reading_remote_location_utc_time() {
        UTCTime utcTime = timeService.utcTime();

        ZonedDateTime utcHawaiiTime = utcTime.atHawaii();
        ZonedDateTime utcLocalTime = utcTime.atLocal();

        assertFalse(TestUtil.formatWithPrecision(utcTime.value(), testProperties.precision()).endsWith("000"));
        assertFalse(TestUtil.formatWithPrecision(utcHawaiiTime.toInstant(), testProperties.precision()).endsWith("000"));
        assertFalse(TestUtil.formatWithPrecision(utcLocalTime.toInstant(), testProperties.precision()).endsWith("000"));
    }

    //------------------------------TAI-------------------------------

    //DEOPSCSW-536: Access parts of TAI date/time in Java and Scala
    //DEOPSCSW-530: SPIKE: Get TAI offset and convert to UTC and Vice Versa
    @Test
    public void should_get_tai_time() {
        TAITime taiTime = timeService.taiTime();
        Instant expectedTaiInstant = Instant.now().plusSeconds(TaiOffset);

        long expectedMillis = expectedTaiInstant.toEpochMilli();

        assertEquals(expectedMillis, taiTime.value().toEpochMilli(), 5);
    }

    //DEOPSCSW-537: Scala and Java API for conversion between TAI and UTC
    //DEOPSCSW-530: SPIKE: Get TAI offset and convert to UTC and Vice Versa
    @Test
    public void should_convert_tai_to_utc() {
        TAITime taiTime = timeService.taiTime();
        UTCTime utcTime = timeService.toUTC(taiTime);

        assertEquals(Duration.between(utcTime.value(), taiTime.value()).getSeconds(), TaiOffset);
    }

    //DEOPSCSW-538: PTP accuracy and precision while reading TAI
    @Test
    public void should_get_maximum_precision_supported_by_system_in_tai() {
        assertFalse(TestUtil.formatWithPrecision(timeService.taiTime().value(), testProperties.precision()).endsWith("000"));
    }

    //DEOPSCSW-541: PTP accuracy and precision while reading remote location time
    @Test
    public void should_get_maximum_precision_while_reading_remote_location_tai_time() {
        TAITime taiTime = timeService.taiTime();

        assertFalse(TestUtil.formatWithPrecision(taiTime.value(), testProperties.precision()).endsWith("000"));
    }

    //------------------------------Scheduling-------------------------------

    //DEOPSCSW-542: Schedule a task to execute in future
    @Test
    public void should_schedule_task_at_start_time() {
        TestProbe testProbe = new TestProbe(untypedSystem);
        String probeMsg = "some message";

        TAITime idealScheduleTime = new TAITime(timeService.taiTime().value().plusSeconds(1));

        Runnable task = () -> testProbe.ref().tell(probeMsg, ActorRef.noSender());

        timeService.scheduleOnce(idealScheduleTime, task);

        manualTime.timePasses(Duration.ofSeconds(1));
        testProbe.expectMsg(probeMsg);
    }

    //DEOPSCSW-544: Schedule a task to be executed repeatedly
    //DEOPSCSW-547: Cancel scheduled timers for periodic tasks
    @Test
    public void should_schedule_a_task_periodically_at_given_interval() {
        List<String> list = new ArrayList<>();

        Cancellable cancellable = timeService.schedulePeriodically(Duration.ofMillis(100), () -> list.add("x"));

        manualTime.timePasses(Duration.ofMillis(500));
        cancellable.cancel();

        assertEquals(list.size(), 6);
    }

    //DEOPSCSW-544: Start a repeating task with initial offset
    //DEOPSCSW-547: Cancel scheduled timers for periodic tasks
    @Test
    public void should_schedule_a_task_periodically_at_given_interval_after_start_time() {
        List<String> list = new ArrayList<>();

        TAITime startTime = new TAITime(timeService.taiTime().value().plusSeconds(1));

        Cancellable cancellable = timeService.schedulePeriodically(startTime, Duration.ofMillis(100), () -> list.add("x"));

        manualTime.timePasses(Duration.ofSeconds(1));
        assertEquals(list.size(), 1);

        manualTime.timePasses(Duration.ofMillis(500));
        cancellable.cancel();
        assertEquals(list.size(), 6);
    }

    //DEOPSCSW-547: Cancel scheduled timers for single scheduled tasks
    @Test
    public void should_cancel_single_scheduled_task(){
        TestProbe testProbe = new TestProbe(untypedSystem);
        String probeMsg = "some message";
        TAITime idealScheduleTime = new TAITime(timeService.taiTime().value().plusSeconds(1));

        Runnable task = () -> testProbe.ref().tell(probeMsg, ActorRef.noSender());

        Cancellable cancellable = timeService.scheduleOnce(idealScheduleTime, task);
        cancellable.cancel();
        testProbe.expectNoMessage(FiniteDuration.apply(500, TimeUnit.MILLISECONDS));
    }

}
